Clickstream logging in Omega

ivmarkp opened the pull request:
This PR enables introduces clickstream logging functionality in Omega as follows:
* An extra $log command in query template added tailored to log data in the required format which includes four fields - Query ID, Query, Hitlist (consists of Documents IDs in hitlitst) and Offset (for pagination). Data would be stored in file named `search.log`
* A new template named clicklog to enable logging of click information (documents that were viewed for the corresponding query (characterised by Doc ID and Query ID)) before redirecting to that document via its URL. Data would be stored in file named `clicks.log`
* A postprocessing script to obtain the final log file named `final.log` using `search.log` and `click.log`. 

Example entries in `final.log`:

```
QueryID	Query	Hits	Offset	Clicks
821f03288846297c2cf43c34766a38f7	book	['45', '36', '14', '54', '42', '52', '2', '3', '15', '3']	0	['0', '0', '0', '2', '0', '0', '0', '0', '0', '0']
098f6bcd4621d373cade4e832627b4f6	test	['45', '36', '14', '54', '42', '52', '2', '3', '15', '3']	0	['0', '0', '0', '0', '1', '0', '0', '0', '0', '0']
```

Example `search.log`:
```
821f03288846297c2cf43c34766a38f7	"book"	45,36,14,54,42,52,2,3,15,32,	0
d41d8cd98f00b204e9800998ecf8427e	""		0
d41d8cd98f00b204e9800998ecf8427e	""		0
098f6bcd4621d373cade4e832627b4f6	"test"	45,36,14,54,42,52,2,3,15,32,	0
```

Example `clicks.log`:
```
821f03288846297c2cf43c34766a38f7	54
821f03288846297c2cf43c34766a38f7	54
098f6bcd4621d373cade4e832627b4f6	42
```

Usage of postprocessing script:

```python
python postprocess.py path/to/search.log path/to/clicks.log
```

ivmarkp commented:
Build failed due to a couple issues in `xapian-core/include/xapian/iterator.h`. I'll rebase this branch on `upstream/master` once that issue is fixed.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,118 @@
+            for row in reader:
+                writer.writerow([row['QueryID'],row['Query']])
+except IOError:
+    print "Could not read file."
```
Errors should be printed to `stderr` not `stdout`, and the program should then `exit()` non-zero. Additionally, this exception could be a failure of write as well as read, so the error message isn't necessarily accurate. It might be easier for now to refactor to three functions for the three blocks, none with exception handling, and then wrap the whole lot in a single exception handler, printing out the exception message (which I believe will include the relevant filename if it's an opening error).

You should also:

 * write python3-compatible code (eg use `print` as a function, not an expression)
 * look into `argparse` so you can provide useful `-h` / `--help` output
 * only act as a program if `__name__ == "__main__"`, which is a common python pattern

Turning the bulk of this into functions would make it possible to write tests for them, as well.

ivmarkp replied:
Will fix all this and will also add `--help` commandline option using `argparse`.

jaylett commented on xapian-applications/omega/templates/query:
```diff
@@ -106,7 +106,7 @@ $if{$field{size},<span title="$html{$field{size}} bytes">Size: <b>$html{$filesiz
 </div>
 </td>
-<td><b><a href="$html{$field{url}}">$html{$or{$field{caption},$field{title},$prettyurl{$field{url}},Untitled}}</a></b><br>
+<td><b><a href="$html{omega.cgi?FMT=$url{clicklog}&QUERYID=$url{$hash{md5,$transform{s/^\s+|\s+$$//g,\1,$query}}}&DOCID=$url{$id}&URL=$url{$field{url}}}">$html{$or{$field{caption},$field{title},$prettyurl{$field{url}},Untitled}}</a></b><br>
```
I believe I asked for this change not to be made to the `query` template, but rather for there to be a new template. You could alternatively do this with a conditional (similar to how we select between `inc/anyalldropbox` and `inc/anyallradio`), so that the default behaviour is not to redirect through `omega.cgi` (which requires installing an extra omegascript template, and an extra server hit — all a bit of a waste to people who don't want the clickstream data at all). Similarly, `search.log` shouldn't be created without changing the default configuration.

Also, `&` in HTML should be escaped.

ivmarkp replied:
> You could alternatively do this with a conditional (similar to how we select between `inc/anyalldropbox` and `inc/anyallradio`)

Sorry, I failed to remember that. Although, I'm not entirely sure how we'd do that. We could select between the two values of `href` tag but what would be the condition for `$if{}` in that case? Could you please explain a bit?

ivmarkp replied:
`Also, & in HTML should be escaped.`

I think since it's inside `$html{}` it automatically gets escaped?

jaylett replied:
Oh, sorry — I missed that it's inside `$html{}`. Yes, that should be fine as is.

jaylett replied:
You do the selection the same way `inc/anyall*` do — accept a default, and let the user change that if they want to switch the other behaviour on. (And then this should be documented — note that it isn't for the previous stuff, so it doesn't really matter where you put it. We should at some point flesh this out into a proper manual for omega, but that's not part of this PR or even project!)

If we wanted it easier to configure, we could use an environment variable (`$env{VAR}`), or possibly using `$opt{}` and `$set{}` (so you'd just set the behaviour you want at the start of the omegascript).

ojwb replied:
This also hard-codes `omega.cgi` - it'd be better to use the same approach as we already do elsewhere in `templates/query` for the form's `action`:

```
<form name=P method=get
action="$html{$or{$env{SCRIPT_NAME},omega}}" target="_top">
```

ivmarkp replied:
@ojwb That's right. I'll make that change.

> I believe I asked for this change not to be made to the query template, but rather for there to be a new template.

@jaylett I might be wrong but I think since such a new template would be much the same as query template except a new `$log` command for `search.log` and this `href` with two conditional values, would it be recommended to add a new template?

jaylett replied:
It will be almost identical yes, which is why I'm suggesting only changing that bit to be a conditional. Have you looked at how we switch between the anyall variants?

jaylett replied:
Since you're going to re-use it, you should compute the hash value first into an opt, and then use it in the two places later.

In earlier discussions, we were going to generate a session id, not a query id. Is there a reason not to do this? It obviously limits the analysis you can do later.

ivmarkp replied:
We can add another field for session id in log files but as I understand it, we'd need to identify different users (using their IP addresses which might raise some privacy concerns) to clearly segregate different sessions duration i.e. the period from the submission of the initial query to the submission of final query. Also, I'm not entirely sure that how it'd be useful in the future from click model training point of view so decided to defer it for now.

jaylett replied:
You don't have to use an IP address, because you don't need to identify users: you can identify the _session_. In this case it's a query session, which starts when you search, and continues while you're conducting the same search. So you want to create a new session id, which continues as you page through search results. The fact that in theory someone could copy that URL and pass it to someone else is probably worth documenting and ignoring (if someone wanted to do anything more "serious", using ip addresses, they can but for privacy reasons they should set that up themselves because only they can understand the consequences under their local legal system).

If you look at CGI parameters such as `xP`, you'll see that there's logic for figuring out if the search has changed. I believe (although I haven't checked quite how it does this) that internally to omega this gets assessed and used to reset things (relevance judgements and the page number, for instance), and would provide a hook for changing the session id.

It's fine to leave this for now if you don't need it for your click modelling, but it's worth making a note of this as a possible future direction.

ojwb replied:
The terms listed in `xP` are compared to those from parsing the current query and there are four cases:

```
// Heuristics:
// * If any terms have been removed, it's a "fresh query" so we discard any
//   relevance judgements
// * If all previous terms are there but more have been added then we keep
//   the relevance judgements, but return the first page of hits
//
// NEW_QUERY entirely new query
// SAME_QUERY unchanged query
// EXTENDED_QUERY new query, but based on the old one
// BAD_QUERY parse error (message in error_msg)
typedef enum { NEW_QUERY, SAME_QUERY, EXTENDED_QUERY, BAD_QUERY } querytype;
```

The criteria for clearing relevance judgements are weaker than those for forcing the first page - we assume if terms are added to an existing query that it's still the same "information need" so documents already marked as relevant should be kept.

jaylett commented on xapian-applications/omega/templates/clicklog:
```diff
@@ -0,0 +1,3 @@
+$httpheader{Status,302 Found}
+$httpheader{Location,$cgi{URL}}
+$log{clicks.log,$cgi{QUERYID}	$cgi{DOCID}}
```
Is there a reason not to include a newline at the end? Some editors will add one on save, which will introduce unwanted changes in future.

I also wonder if it'll be clear to anyone editing that you're using a tab character. (It will depend on whether their editor highlights it, I suspect.) CSV would work fine here since we're dealing with numbers.

ivmarkp replied:
> Is there a reason not to include a newline at the end? Some editors will add one on save, which will introduce unwanted changes in future.

No reason actually. Will add a newline here.

jaylett commented on xapian-applications/omega/templates/query:
```diff
@@ -153,3 +153,4 @@ $map{$cgilist{B},<input type=hidden name=B value="$html{$_}">}
 </body>
 </html>
 $log{query.log}
+$log{search.log,$hash{md5,$transform{s/^\s+|\s+$$//g,\1,$query}}	$csv{$query,1}	$hitlist{$id$.}	$topdoc}
```
Similar comments to the previous logging. There's also something odd (and not necessarily correct) about using `$csv{}` when you're generating a TSV. TSV is flimsily specified, perhaps even more so than CSV :-(

You can avoid the trailing `,` in the hitlist by checking `$hit` against `$sub{$last,1}`.

ivmarkp replied:
> There's also something odd (and not necessarily correct) about using $csv{} when you're generating a TSV. 

Using`$csv{}` here just to quote query strings.

> TSV is flimsily specified, perhaps even more so than CSV :-(

I think we have to use TSV format here because there could be `,` in query strings which will confuse postprocessing script leading to incorrect log records. Is there a better way to specify TSV here?

jaylett replied:
Escaping is how CSV allows `,` in cell values. Using `$csv{}` with a non-CSV file format is potentially dangerous, because the escaping rules won't be identical. (Although as noted, the escaping rules for both CSV and TSV are poorly specified.)

Put it this way: since you also control the input processor (in python), you can ensure that any file format works. So it's better to be clear. Using `$csv{}` with TSV is less clear than using it with CSV.

ojwb replied:
FWIW, if literal tabs are felt to be problematic, then `$chr{9}` can be used.

Why is CSV-style quoting useful when we're using a tab separator?  Is it essentially that we need to quote the string if it contains literal tabs, and `$csv` in "always-quote" mode achieves that?

ivmarkp replied:
> FWIW, if literal tabs are felt to be problematic, then $chr{9} can be used.

That's neat. I'd like to go with this if we can finally agree upon using tab-delimited values?

> Why is CSV-style quoting useful when we're using a tab separator? Is it essentially that we need to quote the string if it contains literal tabs, and $csv in "always-quote" mode achieves that?

Yes, query strings might contain tabs in which case it becomes essential to use quoting. I also checked the final log file entries with and without intentional tabs in query strings and I found that entries with tabs in queries leads were generally erroneous.

jaylett replied:
Okay, I've looked up TSV properly now. Either we should use TSV: strip all tab characters from any field they might appear in (in practice this is just the query string), or we should use CSV and `$csv{}`. [The IANA spec for TSV](https://www.iana.org/assignments/media-types/text/tab-separated-values) (which seems to be the only one that actually exists) is clear:

> Note that fields that contain tabs are not allowable in this encoding.

So quoting is unnecessary in TSV, and may cause other problems (like including the quotes in field values when they're read out using compliant libraries; although compliant libraries are really just things like `.split('\x09')`).

(Technically we aren't really using `text/tab-separated-values` because there's no header. CSV can accommodate this in the MIME type, but that's basically irrelevant for us so we shouldn't worry about it.)

ojwb replied:
With TSV it seems we have to either mangle query strings to eliminate tabs (e.g. changing tabs to spaces) or else define our own variant of TSV with special rules for quoting values containing tabs.  Not being able to preserve queries exactly as typed by the user seems inelegant, while defining our own variant rather ruins the interoperability a standard format ought to give us (because you can't expect existing libraries to handle quoted tabs).

CSV seems a better option as it can at least handle arbitrary query strings without having to mangle them, and omegascript supports CSV escaping already.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,118 @@
+# Postprocess click data files for obtaining the final clickstream log file
+# and generate Query file for Xapian Letor module from it.
```
You can (and probably should) put this bit as a doccomment.

ivmarkp replied:
Okay, will do that.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,118 @@
+
+try:
+    with open(search_log, 'r') as s:
+        with open('log/clickstream.log', 'w+') as f:
```
[Python 2.7](https://docs.python.org/2/reference/compound_stmts.html#the-with-statement) and [Python 3](https://docs.python.org/3.5/reference/compound_stmts.html#the-with-statement) both support multiple context expressions in `with`, separated by commas. Since python 2.6 support ended in 2013, we should be able to use this without problems.

ivmarkp replied:
I actually tried doing that but then this line exceeded 78 characters limit so I changed it to this nested form. Should I change it back if character limit is not an issue?

jaylett replied:
For the two that have two nested `with` the combined line length doesn't exceed 76 characters. The other one you should consider after doing the other changes, because that may change its shape. You should also be able to surround the expression (but not the entire context expression, so the `open()` bit here) in parentheses.

On formatting, I recommend using `flake8` whose defaults are generally good. For modern python code I generally go for a longer line length, such as 100, but only where absolutely necessary.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,118 @@
+                    continue
+                # Remove an extra ',' character at the end of the Hitlist string
+                hits = row[2]
+                hits = hits[:-2]
```
This is removing more than an extra comma. Instead, you can use `hits = hits.rstrip(',')` to strip the last character if it's a comma. (If there's a trailing space, something like `hits.strip().rstrip(',')`.)

ivmarkp replied:
Oh, seems like a typo. Should've been `1` here instead of `2`. I think it will not be required if `,` can be avoided by making use of `$hit` and `$sub{$last,1}` as you pointed out in a comment below. I'll go with that.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,118 @@
+            reader = csv.reader(s, delimiter = '\t')
+            writer = csv.writer(f, delimiter = '\t', quoting = csv.QUOTE_MINIMAL)
+
+            for row in reader:
```
You could do with stating the input file formats and output file formats of this program, since at the moment it's tricky to figure out what it's actually doing.

ivmarkp replied:
Sorry, I'll add more comments explaining the input and output file formats with some example entries.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,118 @@
+                for row1 in reader1:
+                    # Convert column 3 and column 5 values to list type
+                    l2 = literal_eval(row1[2])
+                    l4 = literal_eval(row1[4])
```
That's clever, but I'm not sure why you do it given you have to convert them from strings in the first place. Couldn't you convert them here also? Again, the lack of documentation explaining what the input and output formats are makes this hard to assess.

ivmarkp replied:
This is because of how `csv.writerow()` works. It converts anything passed to it to a string.

So, in line 60, `row` is written into `clickstream.log` as a list of strings which is problematic as `row[2][0]` would give us `[` instead of the first docid that we want and conversion like done before in line 46 wouldn't help much either as we'd then have `[54` for `row[2][0]` but using `literal_eval()` works well for our purposes here.

ivmarkp replied:
This is what types row and column types are just before line 60:
```
row : <class 'list'>
row[0] : <class 'str'>
row[1] : <class 'str'>
row[2] : <class 'list'>
row[3] : <class 'str'>
row[4] : <class 'list'>
```
and after we read from the same `clickstream.log` somewhere after line 78:
```
row1 : <class 'list'>
row1[0] : <class 'str'>
row1[1] : <class 'str'>
row1[2] : <class 'str'>
row1[3] : <class 'str'>
row1[4] : <class 'str'>
```

jaylett replied:
`csv.writerow()` takes an iterable of strings or numbers, so you're taking advantage of the fact that it actually coerces rather than type checking, _and_ that a list coerced to string happens to be parseable as an `eval()` form. If you absolutely had to use an eval form, you should use `repr()` to make the serialised form, but you don't (because it starts life as a serialised form anyway, so just roundtrip through that explicitly) and in fact I remain close to certain you can avoid this entire intermediary approach anyway.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,118 @@
+                reader1 = csv.reader(s, delimiter = '\t')
+
+                # Store the contents of clicks_log in a list object
+                reader2 = list(csv.reader(f, delimiter = '\t'))
```
This is pretty memory hungry. That should at least be pointed out, I think. Also, you could instead read it into a map from queryid to a list of docids, which would make the subsequent searching through it faster.

ivmarkp replied:
Not sure if we need map here. Main purpose of storing in a list is to be able to compare values in each line of `clicks_log` with that of every line of `clickstream.log` in nested loops below.

jaylett replied:
For which `map` will be faster, because it avoids a linear search on every line processed.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,118 @@
+                                count = count + 1
+                                l4[j] = str(count)
+                    # Update "Clicks" list with updated values
+                    row1[4] = l4
```
I suspect you'd find it easier to write out `docid1:count1,docid2:count2,…`, which would also probably make this section easier to read.

ivmarkp replied:
Sure, will make this change.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,118 @@
+                for i in range(len(row[4])):
+                    row[4][i] = str(0)
+
+                writer.writerow(row)
```
So…this ends up with something like:

 * qid
 * query string
 * hitlist, turned into a list then serialised back to string
 * topdoc
 * a list of `0` the same length as hitlist (again, serialised as a string)

However, I'm not sure why you bother writing this to an interim file, rather than collapsing this preprocessing with the subsequent loop.

ivmarkp replied:
> However, I'm not sure why you bother writing this to an interim file, rather than collapsing this preprocessing with the subsequent loop.

At first, I was trying to achieve the same i.e. eliminating the need of having an interim file (`clickstream.log`) but it didn't seem possible to me because there was no way to read and write into `final.log` at the same time.

This is what I tried: opened `search.log` in read mode and `final.log` in write mode to add "Clicks" column to it with all values set to `0` of length same as hitlist and closed it. And, to update those values, we need to be able to read as well as write (row with updated clicks values) which isn't quite possible.

Also, I think it is safer (from data loss perspective) to use independent files to generate `final.log` rather than trying to edit `final.log` itself. What do you think?

jaylett replied:
Why use an intermediate file at all?

ivmarkp replied:
Sorry, you are right; we don't need an intermediate file. Also, as you suggested earlier, I've updated the script to use a mapping of query ids to lists of doc ids using dictionary.

ojwb commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,118 @@
+                # Convert Hitlist from string to list
+                hits = row[2]
+                hits = hits.strip().split(',')
+                row[2] = hits
```
All these literal numbers for indexing into `row` don't make for readable code - I'd suggest setting up some constants which will make the code clearer more self-documenting, e.g. `LOG_QUERY = 1`, `LOG_HITS = 2`, etc and then:

```
if row[LOG_QUERY] == '':
    continue
```

ivmarkp replied:
True. I've introduced these constants with new fixup commit.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+import csv
+import sys
+
+parser = argparse.ArgumentParser(
```
This should be run only if `__name__ == "__main__"`, ie if this is called as a script. If it's imported as a module, we don't want to do this (which will allow you to write tests for the functions).

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+    sys.exit(1)
+
+try:
+    generate_final_log(sys.argv[1], sys.argv[2])
```
I think you mean to pull these two arguments out of `args`, not out of `sys.argv` (which won't get the defaults).

ivmarkp replied:
Yes, need to fix this.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+# Require no less than 3 command line arguments.
+if len(sys.argv) != 3:
+    print("Not enough arguments.", file=sys.stderr)
+    sys.exit(1)
```
This would be taken care of by `argparse` if set up correctly. (Also, all this should be run only when this is run as a script.)

ivmarkp replied:
Yes, this is not required anymore. Will remove this.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+
+try:
+    generate_final_log(sys.argv[1], sys.argv[2])
+    generate_query_file('log/final.log', 'log/query.txt')
```
These two should be configurable on the command line.

ivmarkp replied:
I'll make them the 3rd and 4th command line arguments.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+    QueryID Query   Hits    Offset  Clicks
+    821f03288846297c2cf43c34766a38f7    book    ['45', '36', '14', '54', '42',\
+    '52', '2', '3', '15', '32'] 0   ['45:0', '36:0', '14:0', '54:2', '42:0',\
+    '52:0', '2:0', '3:0', '15:0', '32:0']
```
I still don't understand why you've turned a simple and compact representation of the third field into something considerably more unwieldy.

ivmarkp replied:
This was just to fit these 180+ character lines within the 79 character limit as pointed by flake8. I'll fix these comments.

jaylett replied:
No, the unwieldy representation is how you're storing things, it's nothing to do with wrapping lines. There's no way that `['45', '36, … ]` is better than `45,36,…`. It's just more complex, requiring more work to serialise and deserialise.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+
+    Note: Values in these entries are all of type str.
+    """
+    with open(search_log, 'r') as f, open(clicks_log, 'r') as g:
```
I'd give them better names than `f`, `g`. Maybe `search_f`, `clicks_f` &c?

ivmarkp replied:
I agree. Should have given them better names. Will do that now.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+    with open(search_log, 'r') as f, open(clicks_log, 'r') as g:
+        with open('log/final.log', 'w+') as h:
+            reader1 = csv.reader(f, delimiter='\t')
+            reader2 = csv.reader(g, delimiter='\t')
```
Similarly, `search_reader` or something would be clearer.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+                    # Update did_to_count[did]
+                    click_count = int(did_to_count[did])
+                    click_count += 1
+                    did_to_count[did] = str(click_count)
```
Avoid this moving between `int` and `str` by keeping the count as a number, which will be faster, use less memory, and not confuse python developers who look at this.

ivmarkp replied:
Sure, will do that.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+
+                qid_to_clicks[qid] = did_to_count
+
+            f.seek(0)
```
You haven't read from `f` have you? So you shouldn't need this.

Also, now you're building a lookup map, you can move the processing of the clicks file into its own `with` expression separate to the later processing.

ivmarkp replied:
> You haven't read from f have you? So you shouldn't need this.

Sorry, this is from something else that what I was trying earlier. Will remove this. 

> Also, now you're building a lookup map, you can move the processing of the clicks file into its own with expression separate to the later processing.

Ok, will do that.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+            clicklist = []
+
+            for row in reader1:
+                # Skip rows with empty Query string
```
When would this happen?

ivmarkp replied:
It happens when user clicks on search button without entering any query text.

jaylett replied:
You could avoid that being written to the log in the first place, maybe. Worth commenting if you're going to keep this in, though.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+                hits = hits.strip().split(',')
+                row[HITS] = hits
+
+                clicklist = hits[:]
```
This will produce defaults as `<did>` not `<did>:0` as I think you intend.

ivmarkp replied:
Ah, yes. I'll fix this.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+
+                row.append(clicklist)
+                writer.writerow(row)
+    return
```
No `return` necessary.

ivmarkp replied:
Ok, will remove it.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+                            clicklist[index] = did + ':' + '0'
+
+                row.append(clicklist)
+                writer.writerow(row)
```
This will produce multiple rows with the same `queryid` in the output file, if the same phrase was searched multiple times. Is that what you intend?

ivmarkp replied:
I think duplicates must not be allowed. I'll add a check to keep out duplicates being written in the file. This will subsequently avoid duplicates in `query.txt` as well.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+                        else:
+                            clicklist[index] = did + ':' + '0'
+
+                row.append(clicklist)
```
As I mentioned before, you shouldn't rely on the default coercing behaviour. Convert this explicitly to a string using a defined serialisation that you can reverse later.

ivmarkp replied:
Do you suggest converting `clicklist` to a string and then append it in `row`?

jaylett replied:
Yes, you should explicitly convert to a serialisation of your choosing, rather than relying on default coercing and how that list will happen to be turned into a repr.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,182 @@
+    Example (comma-delimited) entries in query.txt:
+
+    821f03288846297c2cf43c34766a38f7,book
+    098f6bcd4621d373cade4e832627b4f6,test
```
Will duplicates in this file cause problems for the letor tools?

ivmarkp replied:
I'm not familiar with the internals of xapian letor but I think it's best to avoid duplicate entries here.

jaylett commented on xapian-applications/omega/docs/postprocessing.rst:
```diff
@@ -0,0 +1,42 @@
+=========================
+Log Data Postprocessing
+=========================
```
Lines round the header should be the same length.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -68,75 +67,82 @@ def generate_final_log(search_log, clicks_log):
+                # Update did_to_count[did]
+                click_count = did_to_count[did]
+                click_count += 1
+                did_to_count[did] = click_count
```
Just `did_to_count[did] += 1` would be fine.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -68,75 +67,82 @@ def generate_final_log(search_log, clicks_log):
+                did_to_count[did] = 1
+
+            qid_to_clicks[qid] = did_to_count
+            print(qid_to_clicks[qid])
```
You don't want to leave `print()` calls in. (If you want people to be able to see progress in some way, then it should be controllable using something like `--verbose`, probably off by default, and be a little more structured than just printing internal state.)

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -68,75 +67,82 @@ def generate_final_log(search_log, clicks_log):
+        # Add headers to final log file
+        writer.writerow(["QueryID", "Query", "Hits", "Offset", "Clicks"])
+
+        queries = []
```
You should use a `set` for this, which will be much more efficient when checking if a query already exists in it.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -68,75 +67,82 @@ def generate_final_log(search_log, clicks_log):
+                    if did in did_to_count:
+                        clicklist[index] = did + ':' + str(did_to_count[did])
+                    else:
+                        clicklist[index] = did + ':' + '0'
```
Just `+ ':0'`. (And below.)

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -68,75 +67,82 @@ def generate_final_log(search_log, clicks_log):
+                did_to_count = qid_to_clicks[row[QUERYID]]
+                for index, did in enumerate(clicklist):
+                    if did in did_to_count:
+                        clicklist[index] = did + ':' + str(did_to_count[did])
```
I'd probably use `'%s:%i' % (did, did_to_count[did])`. I know some people who'd use `{}`-style string formatting, but it's harder to do that without introducing escaping problems.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -68,75 +67,82 @@ def generate_final_log(search_log, clicks_log):
+                for index, did in enumerate(clicklist):
+                    clicklist[index] = did + ':' + '0'
+
+            print(clicklist)
```
Another rogue `print()` (and a third below).

jaylett commented on xapian-applications/omega/docs/postprocessing.rst:
```diff
@@ -0,0 +1,42 @@
+
+Log data includes a) data logged from the search results page rendered by
+Omega query template and, b) clicks recorded when a document link is
+clicked on the results page.
```
I'd probably use a list of some sort here.

jaylett commented on xapian-applications/omega/docs/postprocessing.rst:
```diff
@@ -0,0 +1,42 @@
+Omega query template and, b) clicks recorded when a document link is
+clicked on the results page.
+
+Further processing of this raw data is handled by the ``postprocess.py``
```
There's probably a better name for this than just "postprocess", but I can't think of one right now.

ojwb replied:
Not sure about the name before the `.`, but I would omit the extension (both from the docs and also from the script itself).  The fact this is written in Python is entirely an implementation detail, and it might end up rewritten in C++ or a different scripting language in the future.  At that point, we have the choice of either dropping the `.py` and breaking any scripts, cronjobs, etc users have which call this program, and making all references to the script name in docs, blog posts, QA sites, etc out of date, or else we have a program called `postprocess.py` which isn't written in Python, which would be strange and confusing.

This is particularly true of scripts which get installed, though I've tried to avoid extensions on scripts in the source tree in general, except when those scripts are inherently specific to a language (as in the bindings) or when there's a hard to avoid reason for needing an extension.

jaylett commented on xapian-applications/omega/docs/postprocessing.rst:
```diff
@@ -0,0 +1,42 @@
+data mining. It also creates a Query file that can be used by Xapian Letor module
+for generating its training files.
+
+The two functions defined in ``postprocess.py`` are:
```
This just makes me itchy to turn all this documentation into a proper sphinx project, so we can use its python documentation support. Let's hold off for now, but can you note that we want to do that? Probably when writing up the next step, which generates data for training letor, would be a good time.

jaylett commented on xapian-applications/omega/docs/postprocessing.rst:
```diff
@@ -0,0 +1,42 @@
+
+The two functions defined in ``postprocess.py`` are:
+
+- ``postprocess.generate_final_log(search_log, clicks_log, final_log)``
```
Perhaps `generate_combined_log()` would be better? Or `combine_logfiles()`, perhaps.

jaylett commented on xapian-applications/omega/docs/postprocessing.rst:
```diff
@@ -0,0 +1,42 @@
+as follows::
+
+    from postprocess import generate_final_log
+    generate_final_log(search_log, clicks_log, final_log)
```
I think this file should contain the documentation of the file formats (you can then be a little more detailed than you would in a docstring). Also, you should link this file from `index.rst` so people can find it :-)

ivmarkp commented:
@jaylett @ojwb I think I have addressed all review comments now with these new commits including the ones that were backlogged for 2-3 days. Please let me know if I missed anything. Thanks.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -78,7 +78,7 @@ def generate_final_log(search_log, clicks_log, final_log):
 
     with open(clicks_log, 'r') as clicks_f:
-        clicks_reader = csv.reader(clicks_f, delimiter='\t')
+        clicks_reader = csv.reader(clicks_f, delimiter=',')
```
`delimiter` defaults to `','`, so you don't have to pass that.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -130,16 +127,18 @@ def generate_final_log(search_log, clicks_log, final_log):
-                        clicklist[index] = did + ':' + '0'
-            else:
+                        clicklist[index] = did + ':0'
+            elif not clicklist:
```
This will never run unless the hitlist is empty, which I don't think is what you want.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -151,7 +151,7 @@ def generate_query_file(final_log, query_file):
     Example (tab-delimited) entries in final.log:
 
-    QueryID	Query	Hits	Offset	Clicks
+    QueryID	Query,Hits,Offset,Clicks
```
Still another tab to nuke here.

ivmarkp commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,187 @@
+    parser.add_argument("clicks_log", type=str, help="Path to the clicks.log file.")
+    parser.add_argument("final_log", type=str, help="Path to save final.log file.")
+    parser.add_argument("query_file", type=str, help="Path to save query.txt file.")
+    parser.add_argument("--test", type=test, help="Run tests for this script.")
```
@jaylett Is this the right way for implementing `--test` to run test function defined in line 159? I'll go ahead and write actual tests in that function if this approach seems fine.

jaylett commented on xapian-applications/omega/docs/postprocessing.rst:
```diff
@@ -42,23 +42,27 @@ as follows::
 ===============
 
-``search_log`` file: Each line in this file contains four fiels as follows:
+**Note:** All files are in CSV format.
```
Again, you should use the `.. note::` admonition here.

jaylett commented on xapian-applications/omega/docs/postprocessing.rst:
```diff
@@ -72,9 +76,9 @@ Log Data Format
+``final_log`` file: Similar to ``search_log`` file but with an additional fifth field as follows:
 
-  1. **QueryID** - an identifier for each query (generated using ``$hash`` OmegaScript command).
+  1. **QueryID** - an identifier for each query.
```
You've removed the reference to `$hash{}` once; I noted it twice but it actually exists three times. Let's clean them all up, please.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -130,8 +116,10 @@ def generate_combined_log(search_log, clicks_log, final_log):
+            else:
                 for index, did in enumerate(clicklist):
+                    if clicklist[index] == '':
+                        continue
```
How can that happen? `clicklist` is generated from `hits`.

ivmarkp replied:
`hits` can be empty sometimes (e.g. in case of a wrongly worded query text). This check is there basically to avoid entries like `:0` under Clicks field.

jaylett replied:
If hits is empty, it will be an empty list, so this won't apply. I still don't see how this can happen.

ivmarkp replied:
It's actually strange to me that the size of `clicklist` for an entry with emply `hits` is `1` when I print it out.

jaylett replied:
An empty string, split on anything, is an empty string. You should detect that when unpacking `hits` and correct for it.

ivmarkp replied:
Okay, got it. Thanks.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -161,12 +149,36 @@ def generate_query_file(final_log, query_file):
+    parser.add_argument("--test", type=test, action='store_true', help="Run tests for this script.")
+    args = parser.parse_args()
+
+if args.test:
```
This needs to live within the main guard, or it'll fail if this is imported as a module. (Although once we're finished that will be impossible.) Then you'll only need once main guard, not two as you have right now.

ivmarkp replied:
Should I move the `try...except` block inside the main guard starting at line 163?

jaylett replied:
That's basically exactly what I said. What are you unclear about?

ivmarkp replied:
I thought you meant to move this `if` block within the main guard above. I see that we only one main guard.

jaylett replied:
I meant to move this `if` within the main guard above. However there are two main guards at the moment.

jaylett commented on xapian-applications/omega/postprocess.py:
```diff
@@ -161,12 +149,36 @@ def generate_query_file(final_log, query_file):
+
+if args.test:
+    test()
+
```
Strictly, there are too many newlines here, although for some reason flake8 doesn't pick that up. (I suspect it may be because of the main guards.)

ivmarkp commented:
@jaylett I see a couple issues with how `--test` is currently being used to run test functions.

Firstly, we need to make the existing positional arguments as optional (by setting their `default` parameter to `None` and `nargs` parameter to `'?'`) to avoid passing file path arguments when all we want is run tests i.e. avoid doing this `python3 postprocess.py --test log/search.log log/clicks.log log/final.log log/query.txt`.

Secondly, even though we make them optional, running tests doesn't produce any helpful ending remark (although, there's always an option to just print out whatever we like) in case of successful tests.

I was wondering if we could use `python3-nose` module to run tests? It can be done with existing functions itself without any further changes. Besides,`--test` flag won't be required in that case. The usage is as follows:

```
python3 -m "nose" postprocess.py
```
which produces a nice output on successful test run:

```
..
----------------------------------------------------------------------
Ran 2 tests in 0.001s

OK
```

jaylett commented:
@ivmarkp you can't use nose for this, because (as per @ojwb's comment) it should be called `postprocess`, not `postprocess.py`, so you can't import it as a module.

jaylett commented:
(Actually, you might be able to use nose directly somehow, but you'd need to drive it yourself, possibly including test discovery, and I've no idea how easy that is. The unittest runner might be easier to make work like that.)

jaylett commented:
The two test functions really should be one, because the second depends on the first having been run. However, you should solve the problem of how to run them first, then worry about that. You could also look at using mocking / patching to avoid having to touch the filesystem, which would make the tests self-contained — otherwise you have the problem that the tests can't easily be run when we install the program, because it won't have the relevant files available to it.

Oh, and we need to install the program as well of course :-)

ojwb commented on xapian-applications/omega/docs/postprocessing.rst:
```diff
@@ -0,0 +1,101 @@
+Further processing of this raw data is handled by the ``postprocess``
+script. This script generates the final clickstream log file from the input
+search and click log files which can be used to train click models for clickstream
+data mining. It also creates a Query file that can be used by Xapian Letor module
```
There doesn't seem to be a reason for capitalising "query" here.

ojwb commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,221 @@
+
+
+def generate_query_file(final_log, query_file):
+    """Generates Query file formatted as per Xapian Letor documentation.
```
Nor here.

ojwb commented on xapian-applications/omega/postprocess.py:
```diff
@@ -0,0 +1,221 @@
+        final_log (string): Path to final.log file.
+        query_file (string): Path to query.txt file.
+
+    Example (tab-delimited) entries in final.log:
```
"tab-delimited" is no longer correct.

ojwb commented on xapian-applications/omega/templates/clicklog:
```diff
@@ -0,0 +1,3 @@
+$httpheader{Status,302 Found}
+$httpheader{Location,$cgi{URL}}
+$log{clicks.log,$cgi{QUERYID}$.$cgi{DOCID}}
```
While the values you pass for `QUERYID` and `DOCID` won't contain characters which need quoting in a CSV file (e.g. commas, double quotes, newlines), someone malicious could call this template passing values for `QUERYID` and `DOCID` which do and that could easily break the format of the log file.  Either the values need vetting before `$log` is used to store them (and bad values discarded), or they need to be quoted with `$csv`.

ojwb commented on xapian-applications/omega/templates/clicklog:
```diff
@@ -0,0 +1,3 @@
+$httpheader{Status,302 Found}
+$httpheader{Location,$cgi{URL}}
```
As currently written, this would allow an attacker to create a URL on the domain hosting omega which forwards to a URL of their choosing.  I wonder if that's problematic - we need to be confident it isn't if we're going to add this as a template installed by default.

If we passed through the `DB` parameter then the `URL` parameter isn't actually needed since the URL can be looked up using the `DOCID` parameter: `$field{url,$cgi{DOCID}}` (though that does mean a little extra work as it has to re-fetch and re-parse the document data).

ivmarkp commented:
@jaylett  I'm not clear on how it'd be possible to mock file paths of files that must contain log entries for generate functions to work on. Both `generate_combined_log()` and `generate_query_file()` take file paths as their arguments. So, I resorted to creating the required files on the fly, use them to test our functions and then delete before exiting the test function.

Also, I've added the `setup` script that will install the program on the user's system.

jaylett commented on xapian-applications/omega/postprocess:
```diff
@@ -0,0 +1,247 @@
+        print(e, file=sys.stderr)
+
+    if args.test:
+        test_functions()
```
This should happen before running `generate_*`, and then exit, to ensure no actual processing happens by mistake.

jaylett commented on xapian-applications/omega/postprocess:
```diff
@@ -0,0 +1,247 @@
+#!/usr/bin/python3
```
This isn't always installed (eg on macOS). `/usr/bin/python` we may be able to assume, or `/usr/bin/env python` might be better. This however means you need to write code that's compatible with both py2 & py3. (It's almost, but not quite, acceptable any more to do py2 only. Big projects are starting the discussion of moving the default over to py3, and some smaller ones already have.)

jaylett commented on xapian-applications/omega/postprocess:
```diff
@@ -0,0 +1,247 @@
+
+"""Postprocesses search and clicks data files.
+
+Generates the final clickstream log file and Query file for Xapian Letor
```
"query"

jaylett commented on xapian-applications/omega/postprocess:
```diff
@@ -0,0 +1,247 @@
+            if i == 1:
+                assert row == '098f6bcd4621d373cade4e832627b4f6,test\n', "Incorrect entry in query.txt"
+
+    # Safely delete test log files.
```
These need to happen as a `finally` in case the earlier code raises an exception (eg by failing an `assert`).

jaylett commented on xapian-applications/omega/postprocess:
```diff
@@ -0,0 +1,247 @@
+
+
+def test_functions():
+    with open('test_search.log', 'w+') as test_search:
```
You shouldn't use hard-coded local file paths, which was partly what I wanted you to change in using mocking/patching. If you want to use temporary files, look at `tempfile` to put them in a run-specific, safe, location.

jaylett commented on xapian-applications/omega/postprocess:
```diff
@@ -0,0 +1,247 @@
+    parser = argparse.ArgumentParser(
+        description='''Postprocess click data files.
+
+        This script generates the final clickstream log file from input search and
```
This string will get formatted in the `--help` output, so it's important you don't include unwanted characters in it. The leading spaces here causes the output to look very odd (there's a surplus newline at the end as well):

```
$ python3 ./xapian-applications/omega/postprocess --help
usage: postprocess [-h] [--test] search_log clicks_log final_log query_file

Postprocess click data files.

        This script generates the final clickstream log file from input search and
        click log files and creates Query file that can be used by Xapian Letor
        module for generating its training files.
        

positional arguments:
  search_log  Path to the search.log file.
  clicks_log  Path to the clicks.log file.
  final_log   Path to save final.log file.
  query_file  Path to save query.txt file.

optional arguments:
  -h, --help  show this help message and exit
  --test      Run tests for this script.
```

jaylett commented on xapian-applications/omega/postprocess:
```diff
@@ -0,0 +1,247 @@
+        description='''Postprocess click data files.
+
+        This script generates the final clickstream log file from input search and
+        click log files and creates Query file that can be used by Xapian Letor
```
"query", also "Xapian's letor module" or "the Xapian letor module" would read better.

ivmarkp commented:
>  if you don't know how to do something, then please ask rather than taking a different approach which may still have problems. Which this one does: you're generating temporary files without managing them properly or putting them where the OS expects you to put temporary files. 

Sorry about that. Should've asked for some elaboration on how to go about using mock. I haven't had a chance to dabble in it earlier so relatively new to this concept.

I have gone through your comments and made the required changes. Please have a look again at the PR. Thanks.

jaylett commented on xapian-applications/omega/postprocess:
```diff
@@ -0,0 +1,246 @@
+        generate_combined_log(args.search_log, args.clicks_log, args.final_log)
+        generate_query_file(args.final_log, args.query_file)
+    except IOError as e:
+        print >> sys.stderr, e
```
What you had before was fine, but you need `from __future__ import print_function`. See [this helpful document](http://python-future.org/compatible_idioms.html#print) about writing code that's compatible with both versions of python.

jaylett commented on xapian-applications/omega/postprocess:
```diff
@@ -0,0 +1,246 @@
+        description='''Postprocess click data files. This script generates the
+        final clickstream log file from input search and click log files and
+        creates query file that can be used by the Xapian Letor module for
+        generating its training files.''')
```
You still have leading spaces in this, which will continue to look weird. (Try it!)

jaylett commented on xapian-applications/omega/postprocess:
```diff
@@ -0,0 +1,246 @@
+    parser.add_argument("search_log", type=str, help="Path to the search.log file.")
+    parser.add_argument("clicks_log", type=str, help="Path to the clicks.log file.")
+    parser.add_argument("final_log", type=str, help="Path to save final.log file.")
+    parser.add_argument("query_file", type=str, help="Path to save query.txt file.")
```
I think for these you should talk about "the search log file" and so forth, rather than "the search.log file". I think it'll be clearer, because it doesn't second guess what people call the files.

jaylett commented on xapian-applications/omega/postprocess:
```diff
@@ -0,0 +1,246 @@
+                    assert row == ['098f6bcd4621d373cade4e832627b4f6','test'], "Incorrect entry in query.txt"
+    finally:
+        # Close all files to safely delete them.
+        test_search.close()
```
These may fail to `.close()`, but also if an exception is raised (say) while writing the first temporary file, the later variables won't exist in scope. So I think you need to wrap these in `try…except` blocks.

ivmarkp commented on xapian-applications/omega/docs/postprocessing.rst:
```diff
@@ -0,0 +1,101 @@
+
+  1. **QueryID** - an identifier for each query.
+  2. **Query** - text of the query (when the search button is clicked on). This field may be empty (e.g. when search button is clicked without any query text entered) in which case **Hits** will also be empty.
+  3. **Hits** - a list of Xapian docid of all documents displayed in the seach results.
```
Fixed it with https://github.com/xapian/xapian/pull/161/commits/b7f6dd6c506198b9ee2595a61a2e965c5d3197d7.

ojwb commented on xapian-applications/omega/templates/query:
```diff
@@ -106,7 +106,7 @@ $if{$field{size},<span title="$html{$field{size}} bytes">Size: <b>$html{$filesiz
 </div>
 </td>
-<td><b><a href="$html{$field{url}}">$html{$or{$field{caption},$field{title},$prettyurl{$field{url}},Untitled}}</a></b><br>
+<td><b><a href="$if{,$html{$or{$env{SCRIPT_NAME},omega.cgi}?FMT=$url{clicklog}&QUERYID=$url{$hash{md5,$transform{s/^\s+|\s+$$//g,\1,$query}}}&DOCID=$url{$id}&URL=$url{$field{url}}},$html{$field{url}}}">$html{$or{$field{caption},$field{title},$prettyurl{$field{url}},Untitled}}</a></b><br>
```
This needs hooking up more usefully than just having an "always false" `$if`.  I'd suggest checking something like `$opt{clicklog}` - then to activate this we can have another template which just does:

```
$set{clicklog,true}$include{query}
```

ivmarkp replied:
I've added `$opt{}` command as suggested. But not sure if you meant to add a new template too as you described or we can just leave that for others to add on their own whenever they want to enable logging?

ojwb replied:
I think it's probably useful to include that template - otherwise we just force everyone to create it for themselves.

ivmarkp replied:
Okay, will include that.

ojwb commented on xapian-applications/omega/templates/query:
```diff
@@ -153,3 +153,4 @@ $map{$cgilist{B},<input type=hidden name=B value="$html{$_}">}
 </body>
 </html>
 $log{query.log}
+$log{search.log,$hash{md5,$transform{s/^\s+|\s+$$//g,\1,$query}}$.$csv{$query,1}$.$csv{$hitlist{$id$if{$ne{$hit,$sub{$last,1}},$.}},1}$.$topdoc}
```
This logs unconditionally.

It also normalises the query string and calculates its hash a second time, which is (a) inefficient and (b) just asking for someone to change one version but not the other and then we end up with the query ids in `search.log` and `clicks.log` not matching.

ivmarkp replied:
I think adding an `$if{}` (based on `$opt{clicklog}`) here should solve the problem of unconditional logging.

However, I don't get how we might end up with non-matching query ids in `search.log` and `clicks.log` because we don't log query ids in `clicks.log`.

ojwb replied:
Your example `clicks.log` given in the python script doc comment shows query ids:

```
821f03288846297c2cf43c34766a38f7,54
```

Is that incorrect?

And if it's not logged there, why do we need to pass the query id to the click logging template?

ivmarkp replied:
No, that is correct. Sorry about the confusion. Not sure why I wrongly said that query IDs are not logged in `clicks.log`.

I think for addressing issue (b), `$cgi{QUERYID}` could be used here instead of calculating the hash a second time.

ojwb replied:
`$cgi{QUERYID}` will only be set in the `clicklog` template (and it is used there).

The issue I'm trying to highlight is with the `query` template - in two different places in this template, the query is munged with a regexp and then an MD5 hash of the result is calculated:

```
$ grep -n -F --colour '$hash{md5,$transform{s/^\s+|\s+$$//g,\1,$query}}' templates/query 
109:<td><b><a href="$if{$opt{clicklog},$html{$or{$env{SCRIPT_NAME},omega.cgi}?FMT=$url{clicklog}&QUERYID=$url{$hash{md5,$transform{s/^\s+|\s+$$//g,\1,$query}}}&DOCID=$url{$id}&URL=$url{$field{url}}},$html{$field{url}}}">$html{$or{$field{caption},$field{title},$prettyurl{$field{url}},Untitled}}</a></b><br>
156:$if{$opt{clicklog},$log{search.log,$hash{md5,$transform{s/^\s+|\s+$$//g,\1,$query}}$.$csv{$query,1}$.$csv{$hitlist{$id$if{$ne{$hit,$sub{$last,1}},$.}},1}$.$topdoc}}
```

If you run it in the terminal yourself you'll get the colour highlighting which makes it clearer.

It'd be better to just calculate this queryid once - the usual approach for this sort of thing is to store it into an option with `$set{queryid,...}` which you can then read with `$opt{queryid}`.

ivmarkp replied:
Thanks for the detailed explanation. I get it now. Will make the required changes.

ojwb commented:
@ivmarkp Ping?

This seems to be fairly close to me, but there's at least my two most recent comments.

ivmarkp commented:
@ojwb Thanks for the reminder. I actually missed earlier notification emails from your last comments.

ojwb commented on xapian-applications/omega/templates/query:
```diff
@@ -106,7 +106,8 @@ $if{$field{size},<span title="$html{$field{size}} bytes">Size: <b>$html{$filesiz
 </div>
 </td>
-<td><b><a href="$if{$opt{clicklog},$html{$or{$env{SCRIPT_NAME},omega.cgi}?FMT=$url{clicklog}&QUERYID=$url{$hash{md5,$transform{s/^\s+|\s+$$//g,\1,$query}}}&DOCID=$url{$id}&URL=$url{$field{url}}},$html{$field{url}}}">$html{$or{$field{caption},$field{title},$prettyurl{$field{url}},Untitled}}</a></b><br>
+$set{queryid,$hash{md5,$transform{s/^\s+|\s+$$//g,\1,$query}}}
```
This should be done **outside** `$hitlist{}` (it doesn't depend on the document, so we can just calculate it once instead of once per document shown).

Also, it ought to be gated by `$opt{clicklog}` - we don't want to waste effort calculating it at all when click logging isn't enabled.

ojwb commented:
Aside from that, this looks OK to me.

@jaylett Are you happy all your points have been addressed?

jaylett commented:
@ivmarkp @ojwb I'm happy with this, yes :)

ojwb commented:
One minor issue is that the fixups don't autosquash, which seems to almost always happen with your PRs.

I think partly this is a symptom of the PRs tending to be a bit large, but especially that there's a lot of iterations needed.  It'd be better to aim for smaller PRs where possible, and to take a bit more time up front looking over the changes for problems before opening the PR.  That's likely to actually get code merged faster, and is better practice - outside of GSoC, you'll tend to find maintainers less patient of patches which need this many iterations.

Also, thinking about the commit boundaries and how the fixups will apply might help - the reason they don't apply cleanly is typically due to a commit between the fixup and the commit it fixes up which changes the same code - that's a bit of a sign that the commit boundaries perhaps aren't in the right places.

Anyway, I've concluded in this case, it's probably simplest to just to squash to one commit - the clashes are fiddly to resolve, and while it could be split, it isn't terrible as a single commit.

However, while looking at this, I noticed a few remaining issues:

The clicklog template gets passed parameters:

```
FMT=$url{clicklog}&QUERYID=$opt{queryid}&DOCID=$url{$id}&URL=$url{$field{url}}},$html{$field{url}}}
```

Firstly, I notice `URL` is unused by the `clicklog` template.  And `$url{clicklog}` is pointlessly escaping a literal value (`clicklog`) which doesn't need escaping.  But more problematic is that this doesn't pass `DB` which means that `$field{url,$cgi{DOCID}}` in the `clicklog` template will assume `DB=default` - that's OK if that's actually what we're searching, but won't work in general.  Either there's no `default` DB and the redirect will fail to find the URL, or it's a different database and we'll either send the user to an arbitrary URL, or give an error if there's no document with that particular docid.

Also, the `clicklog` template is only put on one of the clickable URLs on each hit, which means we'll only log when the user clicks the link from the title, not the link from the URL text:

```
<td><b><a href="$if{$opt{clicklog},$html{$or{$env{SCRIPT_NAME},omega.cgi}?FMT=$url{clicklog}&QUERYID=$opt{queryid}&DOCID=$url{$id}&URL=$url{$field{url}}},$html{$field{url}}}">$html{$or{$field{caption},$field{title},$prettyurl{$field{url}
},Untitled}}</a></b><br>
<small>$snippet{$field{sample}}</small><br>
<a href="$html{$field{url}}">$html{$field{url}}</a><br>
```

The link at the end here goes straight to the URL.

You can calculate the URL once per match and stick it in an option with `$set`, which would also make this part of the template easier to follow.

Sorry for not spotting these before.

ivmarkp commented:
>  I think partly this is a symptom of the PRs tending to be a bit large, but especially that there's a lot of iterations needed. It'd be better to aim for smaller PRs where possible, and to take a bit more time up front looking over the changes for problems before opening the PR. That's likely to actually get code merged faster, and is better practice - outside of GSoC, you'll tend to find maintainers less patient of patches which need this many iterations.

Sorry for too many iterations on this PR. I really appreciate the time everyone spent on reviews. I'll try to keep the number of iterations down to minimum in the future. Thanks for your advice. 

> However, while looking at this, I noticed a few remaining issues:

I've addressed both the issues that you pointed out. Hopefully, this won't go through any more iterations.

ojwb commented on xapian-applications/omega/templates/query:
```diff
@@ -88,6 +88,7 @@ $list{$map{$queryterms,$list{$html{$uniq{$unstem{$_}}},<b>,</b>/<b>,</b>}:&nbsp;
 <table>
 $if{$opt{clicklog},$set{queryid,$hash{md5,$transform{s/^\s+|\s+$$//g,\1,$query}}}}
 $hitlist{<tr><td valign=top>
+$if{$opt{clicklog},$set{docurl,$field{url}}}
```
Um, I meant put the URL that the click should lead in an option - i.e. `$field{url}` in the standard case, and the URL which bounces the user via omega with the `clicklog` template in the click logging case.  As is, you've still got two copies of the code to build the `clicklog` URL in the template, which is harder to maintain (it's all too easy to only update one version), and which makes that part of the template harder to read.

Just changing `$field{url}` to `$opt{docurl}` makes little sense - the latter is no clearer, and there's also the extra code to set it.

ojwb commented on xapian-applications/omega/templates/query:
```diff
@@ -108,9 +108,9 @@ $if{$field{size},<span title="$html{$field{size}} bytes">Size: <b>$html{$filesiz
+<td><b><a href="$if{$opt{clicklog},$opt{clicklogurl},$html{$field{url}}}">$html{$or{$field{caption},$field{title},$prettyurl{$field{url}},Untitled}}</a></b><br>
 <small>$snippet{$field{sample}}</small><br>
-<a href="$if{$opt{clicklog},$html{$or{$env{SCRIPT_NAME},omega.cgi}?FMT=clicklog&DB=$url{$dbname}&QUERYID=$opt{queryid}&DOCID=$url{$id}&URL=$url{$opt{docurl}}},$html{$opt{docurl}}}">$html{$opt{docurl}}</a><br>
+<a href="$if{$opt{clicklog},$opt{clicklogurl},$html{$field{url}}}">$html{$field{url}}</a><br>
```
ARGH!  This is still unhelpfully duplicating code - we still have two copies of this code to decide which URL to insert:

```
$if{$opt{clicklog},$opt{clicklogurl},$html{$field{url}}}
```

What I'm suggesting should go in the option is the URL which we want in both these links.  So the second use should look like:

```
<a href="$html{$opt{clickurl}}">$html{$field{url}}</a>
```

Here `$opt{clickurl}` is the URL the user gets to click on and `$field{url}` which is the URL of the resource itself.  In the default case the two are the same, but in the "clicklog" case the "clickurl" takes us via the redirect.

I'd suggest keeping the `$html{...}` escaping step at the point of use, as that gives a cleaner separation between unescaped and escaped data.  If some data is already escaped and some isn't, it's harder to follow where escaping is needed, which means more of a breeding ground for bugs.

ivmarkp commented on xapian-applications/omega/templates/query:
```diff
@@ -108,9 +108,9 @@ $if{$field{size},<span title="$html{$field{size}} bytes">Size: <b>$html{$filesiz
+<td><b><a href="$html{$opt{clickurl}}">$html{$or{$field{caption},$field{title},$prettyurl{$field{url}},Untitled}}</a></b><br>
 <small>$snippet{$field{sample}}</small><br>
-<a href="$if{$opt{clicklog},$opt{clicklogurl},$html{$field{url}}}">$html{$field{url}}</a><br>
+<a href="$html{$opt{clickurl}}">$prettyurl{$html{$field{url}}}</a><br>
```
I think html escaping isn't required in `$prettyurl{$html{$field{url}}}` instead we probably just want `$prettyurl{$field{url}}` here?

jaylett replied:
If the URL has `'&'` in, it will need escaping for HTML, so `$html{}` is required.

ojwb replied:
`$html` is needed, though really the escaping should be the done **last** (so the outermost operation) - that way it's clear the value is properly escaped (assuming `$html` does its job correctly, which is something that can be tested in isolation).

If you use `$prettyurl{$html{...}}` the reader can't be sure that the output is properly escaped for HTML without looking carefully at what `$prettyurl` does and thinking through how that can interact with HTML-escaped data.

However, you're actually adding a `$prettyurl` which wasn't here before.  It's really a bug that it isn't already used here, and it would be good to fix that bug, but make this change as part of a commit to add clickstream logging is unhelpful.  Each commit should do one thing.  Prettifying the URL here is not connected with clickstream logging, so should be a separate commit.

In particular, adding this missing `$prettyurl` is a candidate for backporting to 1.4.x, but clickstream logging isn't something that's likely to be appropriate to backport.

ojwb replied:
I've pushed the `$prettyurl` change separately, as it'll be simpler to backport if we make it first.  Just drop it from here, and we'll pick it up when this gets merged.

ivmarkp replied:
Okay, done; dropped the `$prettyurl` change.

jaylett commented on xapian-applications/omega/templates/query:
```diff
@@ -88,7 +88,7 @@ $list{$map{$queryterms,$list{$html{$uniq{$unstem{$_}}},<b>,</b>/<b>,</b>}:&nbsp;
 $if{$opt{clicklog},$set{queryid,$hash{md5,$transform{s/^\s+|\s+$$//g,\1,$query}}}}
 $hitlist{<tr><td valign=top>
-$if{$opt{clicklog},$set{clicklogurl,$html{$or{$env{SCRIPT_NAME},omega.cgi}?FMT=clicklog&DB=$url{$dbname}&QUERYID=$opt{queryid}&DOCID=$url{$id}&URL=$url{$field{url}}}}}
+$if{$opt{clicklog},$set{clickurl,$or{$env{SCRIPT_NAME},omega.cgi}?FMT=clicklog&DB=$url{$dbname}&QUERYID=$opt{queryid}&DOCID=$url{$id}&URL=$url{$field{url}}},$set{clickurl,$html{$field{url}}}}
```
Would this be clearer as `$set{clickurl,$if{$opt{clicklog},VIA-OMEGA,$field{url}}`?

In particular, it might make it clearer that the different legs here aren't equivalent: one still wraps in `$html{}`.

ivmarkp replied:
Yes, that's true. Will make the required change here.

ojwb commented on xapian-applications/omega/templates/query:
```diff
@@ -88,7 +88,7 @@ $list{$map{$queryterms,$list{$html{$uniq{$unstem{$_}}},<b>,</b>/<b>,</b>}:&nbsp;
 <table>
 $hitlist{<tr><td valign=top>
-$set{clickurl,$if{$opt{clicklog},$or{$env{SCRIPT_NAME},omega.cgi}?FMT=clicklog&DB=$url{$dbname}&QUERYID=$opt{queryid}&DOCID=$url{$id}&URL=$url{$field{url}},$field{url}}}
+$set{clickurl,$if{$opt{clicklog},$or{$env{SCRIPT_NAME},omega.cgi}?FMT=clicklog&DB=$url{$dbname}&QUERYID=$opt{queryid}&DOCID=$url{$id}&URL=$url{$field{url}},$html{$field{url}}}}
```
This undoes the (correct) fix you made two commits earlier.

As a rule of thumb, escape with `$html` at the point of use, and only there.

If you escape twice, (as happens after this change in the non-clicklog case) you'll break stuff.  For example, a URL with `&` in will get escaped to `&amp;amp;`, which the browser will unescape to `&amp;` when the result of unescaping needs to be just `&`.

ivmarkp replied:
Ah, I see. I didn't encounter anything unusual with non-clicklog case when I tested these changes on the browser because no URL in my case has `&` in it. Thanks for the explanation; will fix this.

ojwb commented on xapian-applications/omega/templates/query:
```diff
@@ -86,7 +87,6 @@ $if{$not{$msizeexact},
 $list{$map{$queryterms,$list{$html{$uniq{$unstem{$_}}},<b>,</b>/<b>,</b>}:&nbsp;$nice{$freq{$_}}},Term frequencies: ,$. ,}
 <br><small>Search took $time seconds</small>
 <table>
-$if{$opt{clicklog},$set{queryid,$hash{md5,$transform{s/^\s+|\s+$$//g,\1,$query}}}}
```
What was the reason for moving this earlier?

It seems it was OK where it was to me, and it'll now get executed in cases when it'll never get used (for example when there's an error).

ivmarkp replied:
Because for queries leading to zero hits (e.g. lorem ipsum) `QUERYID` was not being logged (or logged blank ) in `search.log`.

ojwb commented:
Because for queries leading to zero hits (e.g. lorem ipsum) QUERYID
was not being logged (or logged blank ) in search.log.


It seems to me such queries aren't useful to log here - when there are no
results there can't be any clicks on results...

ivmarkp commented:
> It seems to me such queries aren't useful to log here - when there are no results there can't be any clicks on results...

That's true. I'll add a check in `postprocess` for empty hitlist in search log entries and revert this change.

ivmarkp commented on xapian-applications/omega/postprocess:
```diff
@@ -0,0 +1,262 @@
+    parser.add_argument("clicks_log", type=str, help="Path to the clicks.log file.")
+    parser.add_argument("final_log", type=str, help="Path to save final.log file.")
+    parser.add_argument("query_file", type=str, help="Path to save query.txt file.")
+    parser.add_argument("--test", help="Run tests for this script.", action='store_true')
```
@jaylett Currently, to run tests, we need to pass values for each of the above four positional arguments (e.g. `./postprocess --test log/search.log log/clicks.log log/final.log log/query.txt`) but the test function doesn't really use them. Probably we need to add `nargs="?"` parameter to these four positional args to be able to run tests without having to pass unnecessary args (i.e. `./postprocess --test`)?

jaylett replied:
Given `--test` is internal only, I think it's fine to pass through dummy args in this case. That's better than having to write more code to cope with the non-test case and ensure all the args have been passed correctly.

ivmarkp commented on xapian-applications/omega/postprocess:
```diff
@@ -201,12 +202,6 @@ def test_functions():
-        generate_query_file(test_final.name, test_query.name)
-
-        test_query.flush()
-
```
Made this change to resolve `test_query` variable scope error (`UnboundLocalError: local variable test_query referenced before assignment`) at `test_query.flush()` being reported when one of the assert statements fails for final log entries above.

ojwb commented:
> > It seems to me such queries aren't useful to log here - when there are no results there can't be any clicks on results...
>
> That's true. I'll add a check in `postprocess` for empty hitlist in search log entries and revert this change.

That still seems the wrong place to address this.

What's the point in logging entries we already know we'll never need at the point we log them?

By logging them, we bloat `query.log`.  And that means more disk I/O on the search server writing those entries, more disk space to store them, and then more disk I/O and CPU to read and discard them when we process the logs.  Not much per entry, but we want this system to be able to deal with large volumes of searches (especially since that's where it's going to work best).

I think this is really a sign that the logging to `query.log` happens in the wrong place - we should only log the query if there are any matches.

It would actually be more robust to log to `query.log` **before** we write out any click-through links which use the query-id.  Currently we may see clicks which have no `query.log` entry if omega exits for some reason between writing out the click through links and logging the query.  That's something of a corner case though, and isn't a big deal so long as the `postprocess` script does something sensible with such clicks rather than crashing out.  But if we're moving this logging, perhaps we should move it to also address this.

ojwb commented on xapian-applications/omega/templates/activatelog:
```diff
@@ -0,0 +1,2 @@
+$set{clicklog,true}
+$include{query}
```
Sorry - I missed this before, but there shouldn't be a newline (or anything which gets output to the browser) before `$include{query}`.  If there is that means that the `$httpheader{Content-Type,text/html; charset=utf-8}` at the start of the `query` template can't work, because `$httpheader` can only work before anything has been output.

ojwb replied:
I checked the code, and it seems that currently this will actually work as-is.  However, we explicitly document it doesn't:

> Headers must be specified before any other
> output from the OmegaScript template - any `$httpheader{}`
> commands found later in the template will be silently ignored.

So it would be prudent not to rely on the current behaviour.

ojwb commented on xapian-applications/omega/templates/clicklog:
```diff
@@ -0,0 +1,3 @@
+$httpheader{Status,302 Found}
+$httpheader{Location,$field{url,$cgi{DOCID}}}
```
Similarly, we shouldn't have a blank link between the two `$httpheader` commands here.

ivmarkp commented:
> It would actually be more robust to log to search.log before we write out any click-through links which use the query-id.

Sorry, I didn't get this part. Can you please elaborate a bit?

ojwb commented:
> Can you please elaborate a bit?

In the middle of the `query` template, we're writing out HTML with links containing the generated query id, but only at the end of the template do we actually log that query id to `search.log`.

If we're sending the page to the client (i.e. the user's web browser) as we generate it [*], then that means there's a short interval of time during which the CGI could die (crash, be killed, etc) and we'll not have logged the query id, but will already have sent HTML with links containing it.  If an end user clicks on one of those links, we'll have a `click.log` entry containing a query id which isn't in `search.log`.

Looking at `postprocess` it seems that will effectively just ignore click data when there's no `search.log` entry for the given query id, which is probably the best it can do at that point, but it'd be better to avoid this happening.

So what I'm suggesting is that in the "there are some matches" part of the template we do:

 * generate query id
 * log query id to `search.log`
 * for each match:
   * show match with link to `clicklog` template using query id

That way we should avoid any possibility of such "orphaned" clicks.

---

[*] Looking at the omega code, it looks like we actually currently end up buffering the whole page until the template has completed, which isn't what I thought happened, and is far from ideal as it means the perceived page load time will be higher as the browser can't start to render the partial page (users feel the page has loaded faster if they see it start to appear on the screen sooner, even if the total time is the same).  It likely also means the actual page load time will be a bit longer.  So this is really an implementation detail that's liable to change.

ivmarkp commented:
@ojwb Thanks for the explanation. I've moved the logging command that logs to `search.log` such that now the queries with zero hits are not logged in `search.log`.
